From 622fa22c37a89fedb0de50d8576de3e0e6a96f60 Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Wed, 3 Jul 2013 22:43:33 +0800
Subject: [PATCH 13/14] set device failure event with m0_poolmach. m0repair and
 snc cm only change device state from FAILED -> SNS_REPAIRING->SNS_REPAIRED.

---
 m0t1fs/linux_kernel/st/m0t1fs_sns_repair.sh | 22 ++++++++++++++++++++++
 pool/pool_trigger.c                         | 29 +++++++++++++++++++++++++++--
 sns/cm/cm.c                                 | 13 +------------
 3 files changed, 50 insertions(+), 14 deletions(-)

diff --git a/m0t1fs/linux_kernel/st/m0t1fs_sns_repair.sh b/m0t1fs/linux_kernel/st/m0t1fs_sns_repair.sh
index 6344f37..7d84712 100755
--- a/m0t1fs/linux_kernel/st/m0t1fs_sns_repair.sh
+++ b/m0t1fs/linux_kernel/st/m0t1fs_sns_repair.sh
@@ -49,6 +49,17 @@ sns_repair_test()
 		IOSEP="$IOSEP -S ${lnet_nid}:${EP[$i]}"
 	done
 
+####### Set Failure device
+	poolmach="$MERO_CORE_ROOT/pool/m0poolmach -O Set -T device -I $fail_device -s 1
+                         -C ${lnet_nid}:${SNS_CLI_EP} $IOSEP"
+	echo $poolmach
+
+	if ! $poolmach ; then
+		echo "m0poolmach failed"
+		unmount_and_clean &>> $MERO_TEST_LOGFILE
+		return 1
+	fi
+
 	trigger="$MERO_CORE_ROOT/sns/cm/st/m0repair -O 2 -F $fail_device
                          -C ${lnet_nid}:${SNS_CLI_EP} $IOSEP"
 	echo $trigger
@@ -62,6 +73,17 @@ sns_repair_test()
 		rc=$?
 	fi
 
+####### Query device state
+	poolmach="$MERO_CORE_ROOT/pool/m0poolmach -O Query -T device -I $fail_device
+                         -C ${lnet_nid}:${SNS_CLI_EP} $IOSEP"
+	echo $poolmach
+
+	if ! $poolmach ; then
+		echo "m0poolmach failed"
+		unmount_and_clean &>> $MERO_TEST_LOGFILE
+		return 1
+	fi
+
 	unmount_and_clean &>> $MERO_TEST_LOGFILE
 
 	return $rc
diff --git a/pool/pool_trigger.c b/pool/pool_trigger.c
index 1d69dd5..9022a56 100644
--- a/pool/pool_trigger.c
+++ b/pool/pool_trigger.c
@@ -51,6 +51,16 @@ enum {
 	MAX_SERVERS        = 1024
 };
 
+char *poolmach_state_name[] = {
+	[M0_PNDS_ONLINE]          = "Online",
+	[M0_PNDS_FAILED]          = "Failed",
+	[M0_PNDS_OFFLINE]         = "Offline",
+	[M0_PNDS_SNS_REPAIRING]   = "SNS Repairing",
+	[M0_PNDS_SNS_REPAIRED]    = "SNS Repaired",
+	[M0_PNDS_SNS_REBALANCING] = "SNS Rebalancing",
+	[M0_PNDS_SNS_REBALANCED]  = "SNS Rebalanced",
+};
+
 const char *cl_ep_addr;
 const char *srv_ep_addr[MAX_SERVERS];
 const char *dbname = "sr_cdb";
@@ -344,14 +354,29 @@ int main(int argc, char *argv[])
 				       0);
 #else
 		rc = m0_rpc_client_call(req, session, NULL, 0);
+		if (rc != 0) {
+			m0_fop_put(req);
+			return rc;
+		}
+
 		if (op[0] == 'Q' || op[0] == 'q') {
 			struct m0_fop_poolmach_query_rep *query_fop_rep;
 			struct m0_fop *rep;
+			char          *name;
+			uint32_t       state;
 			rep = m0_rpc_item_to_fop(req->f_item.ri_reply);
 			query_fop_rep = m0_fop_data(rep);
-			fprintf(stderr, "Query got reply: rc = %d, state=%d\n",
+			state = query_fop_rep->fpq_state;
+			name = (state >= 0 &&
+				state < ARRAY_SIZE(poolmach_state_name))?
+					poolmach_state_name[state]:
+					"N/A";
+
+			fprintf(stderr, "Query got reply: rc = %d, "
+					"state = %d (%s)\n",
 					(int)query_fop_rep->fpq_rc,
-					(int)query_fop_rep->fpq_state);
+					(int)state,
+					name);
 		} else {
 			struct m0_fop_poolmach_set_rep *set_fop_rep;
 			struct m0_fop *rep;
diff --git a/sns/cm/cm.c b/sns/cm/cm.c
index f1071e4..3f03b1e 100644
--- a/sns/cm/cm.c
+++ b/sns/cm/cm.c
@@ -562,10 +562,7 @@ static int pm_state(struct m0_sns_cm *scm)
 static int cm_ready(struct m0_cm *cm)
 {
 	struct m0_sns_cm      *scm = cm2sns(cm);
-	struct m0_dbenv       *dbenv = scm->sc_it.si_dbenv;
 	int                    bufs_nr;
-	int                    rc = 0;
-	int                    i;
 
 	M0_ENTRY("cm: %p", cm);
 	M0_PRE(M0_IN(scm->sc_op, (SNS_REPAIR, SNS_REBALANCE)));
@@ -594,18 +591,10 @@ static int cm_ready(struct m0_cm *cm)
 		 * handle multiple failures.
 		 */
 		scm->sc_failures_nr = 1;
-		for (i = 0; i < scm->sc_failures_nr; ++i) {
-			rc = pm_event_setup_and_post(dbenv, cm->cm_pm,
-						     M0_POOL_DEVICE,
-						     scm->sc_it.si_fdata[i],
-						     M0_PNDS_FAILED);
-			if (rc != 0)
-				return rc;
-		}
 	}
 
 	M0_LEAVE();
-	return rc;
+	return 0;
 }
 
 static int cm_start(struct m0_cm *cm)
-- 
1.8.3.2

